src/contracts/deposit/deposit-tree/deposit-tree.ts

```diff
@@ -0,0 +1,138 @@
+import { fromHexString } from '@chainsafe/ssz';
+import { parseLittleEndian64, toLittleEndian64 } from '../deposit.utils';
+import { DepositData } from 'bls/bls.containers';
+import { NodeData } from '../interfaces';
+
+export class DepositTree {
+  static DEPOSIT_CONTRACT_TREE_DEPTH = 32;
+  static ZERO_HASH = fromHexString(
+    '0x0000000000000000000000000000000000000000000000000000000000000000',
+  );
+  zeroHashes: Uint8Array[] = new Array(DepositTree.DEPOSIT_CONTRACT_TREE_DEPTH);
+  branch: Uint8Array[] = [];
+  nodeCount = 0;
+
+  constructor() {
+    this.formZeroHashes();
+  }
+
+  /**
+   * Initializes the zero hashes used in the tree.
+   */
+  private formZeroHashes() {
+    this.zeroHashes[0] = DepositTree.ZERO_HASH;
+    for (
+      let height = 0;
+      height < DepositTree.DEPOSIT_CONTRACT_TREE_DEPTH - 1;
+      height++
+    ) {
+      this.zeroHashes[height + 1] = digest2Bytes32(
+        this.zeroHashes[height],
+        this.zeroHashes[height],
+      );
+    }
+  }
+
+  /**
+   * Forms the branch of the tree needed to update the root when a new node is inserted.
+   * @param {Uint8Array} node - The node's data to be inserted.
+   * @param {number} depositCount - The sequential index of the deposit, representing the total deposits.
+   * @returns {Uint8Array[] | undefined} The updated branch of the tree after inserting the node.
+   */
+  private formBranch(
+    node: Uint8Array,
+    depositCount: number,
+  ): Uint8Array[] | undefined {
+    let size = depositCount;
+    for (
+      let height = 0;
+      height < DepositTree.DEPOSIT_CONTRACT_TREE_DEPTH;
+      height++
+    ) {
+      if ((size & 1) == 1) {
+        this.branch[height] = node;
+        return this.branch;
+      }
+
+      node = digest2Bytes32(this.branch[height], node);
+
+      size /= 2;
+    }
+  }
+
+  /**
+   * Inserts a new deposit into the tree using detailed node data.
+   * @param {NodeData} nodeData - The detailed data of the deposit to be inserted.
+   */
+  public insert(nodeData: NodeData) {
+    const node = DepositTree.formDepositNode(nodeData);
+    this.nodeCount++;
+    this.formBranch(node, this.nodeCount);
+  }
+
+  /**
+   * Inserts a new node into the tree using already computed node hash.
+   * @param {Uint8Array} node - The node's hash to be inserted.
+   */
+  public insertNode(node: Uint8Array) {
+    this.nodeCount++;
+    this.formBranch(node, this.nodeCount);
+  }
+
+  /**
+   * Computes and returns the root hash of the deposit tree.
+   * @returns {string} The computed root hash of the tree.
+   */
+  public getRoot() {
+    let node = DepositTree.ZERO_HASH;
+    let size = this.nodeCount;
+    for (
+      let height = 0;
+      height < DepositTree.DEPOSIT_CONTRACT_TREE_DEPTH;
+      height++
+    ) {
+      if ((size & 1) == 1) {
+        node = digest2Bytes32(this.branch[height], node);
+      } else {
+        node = digest2Bytes32(node, this.zeroHashes[height]);
+      }
+      size /= 2;
```

Comment:
will it work properly if size is odd? or we will get here an odd number

Reply:
Using `size /= 2` is not a mistake. In JavaScript, when performing bitwise operations like `& 1`, floating-point numbers are implicitly converted to integers, discarding the fractional part. This ensures the algorithm works correctly and matches the logic of a Solidity smart contract. Solidity does not have floating-point numbers, and all division is performed as integer division, rounding down the result.